### PR TITLE
Implement non-boolean bitwise complement

### LINKED ISSUE #18788: Support bitwise complement (non-boolean Expression.Not)
*DISCLAIMER*: I got sucked into this just as I was about to go home, so technically this is a project done on my free personal time!

Our codebase mistakenly assumes in various places that ExpressionType.Not is boolean. Make sure bitwise complement (non-boolean ExpressionType.Not) is properly translated.

Supported by SqlServer, Sqlite, PostgreSQL, even Cosmos...

### PR DESCRIPTION
**Pull request by roji**
Fixes #18788

### CODE REVIEW

File: src/EFCore.Cosmos/Extensions/Internal/RelationalExpressionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Internal
+{
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public static class CosmosExpressionExtensions
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Extensions/Internal/RelationalExpressionExtensions.cs]: The file should be named `CosmosExpressionExtensions.cs`

File: src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs [hunk 2]
```diff
@@ -107,7 +108,8 @@ public virtual SqlExpression ApplyTypeMapping(SqlExpression sqlExpression, CoreT
             {
                 case ExpressionType.Equal:
                 case ExpressionType.NotEqual:
-                case ExpressionType.Not:
+                case ExpressionType.Not
+                    when sqlUnaryExpression.IsLogicalNot():
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs]: And braces to visually separate the body.

Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs]: I'd prefer to just put the when clause on the same line as the case... But I guess we don't do that?

Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Internal/SqlExpressionFactory.cs]: R# doesn't allow that

**Pull request merged**